tv_apps/browser/js/datasync/sync_bookmarks.js

```diff
@@ -11,23 +11,38 @@
 var SyncBookmark = (function () {
   var syncDataStore;
 
-  function handleTask(task) {
+  function updateWrapper(data) {
     return new Promise(resolve => {
-      var data;
-      if (task.data) {
-        data = task.data.fxsyncPayload;
-        data.timestamp = task.data.last_modified;
-      }
-      switch(task.operation) {
-      case 'update':
-        SyncBrowserDB.updateBookmark(data, () => {
+      if (data.deleted) {
+        console.log(data); // XXX
+        SyncBrowserDB.removeBookmark(data.id, () => {
           resolve();
         });
-        break;
-      case 'add':
-        SyncBrowserDB.addBookmark(data, () => {
+      } else {
+        SyncBrowserDB.updateBookmark(data, () => {
           resolve();
         });
+      }
+    });
+  }
+
+  function handleTask(task) {
+    return new Promise((resolve, reject) => {
+      var data = [];
+      if (task.data && task.data.fxsyncRecords) {
+        for (var key in task.data.fxsyncRecords) {
+          data.push(task.data.fxsyncRecords[key]);
+        }
+      }
+      switch(task.operation) {
+      case 'update':
+      case 'add':
+        //Promise.all(data.map(updateWrapper)).then(resolve);
+        data.reduce((reduced, current) => {
+          return reduced.then(() => {
+            return updateWrapper(current);
+          });
+        }, Promise.resolve()).then(resolve);
```

Comment:
Nicer code pattern would be `Promise.all(data.map(fxSyncRecord => updateWrapper(fxSyncRecord))`